File: algorithms/code/leetcode/lc102_binary_tree_level_order_traversal/lc102_binary_tree_level_order_traversal.py
#!/usr/bin/env python3
# https://leetcode.com/problems/binary-tree-level-order-traversal/

import unittest
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    class AnnotatedNode:
        def __init__(self, node, depth):
            self.node = node
            self.depth = depth

    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result: List[List[int]] = []
        queue = []
        if root is not None:
            queue.append(self.AnnotatedNode(root, 1))
        depth = 0
        line: List[int] = None
        while queue:
            annotated = queue.pop(0)
            if depth != annotated.depth:
                depth = annotated.depth
                line = []
                result.append(line)
            # if line is not None:
            line.append(annotated.node.val)
            if annotated.node.left is not None:
                queue.append(self.AnnotatedNode(
                    annotated.node.left, annotated.depth + 1))
            if annotated.node.right is not None:
                queue.append(self.AnnotatedNode(
                    annotated.node.right, annotated.depth + 1))
        return result


class TestCode(unittest.TestCase):
    def test_example(self):
        n3 = TreeNode(3)
        n7 = TreeNode(7)
        n9 = TreeNode(9)
        n15 = TreeNode(15)
        n20 = TreeNode(20)
        n3.left = n9
        n3.right = n20
        n20.left = n15
        n20.right = n7
        expected = [[3], [9, 20], [15, 7]]
        result = Solution().levelOrder(n3)
        self.assertEqual(len(expected), len(result))
        for i, _ in enumerate(expected):
            self.assertEqual(len(expected[i]), len(result[i]))
            for j, _ in enumerate(expected[i]):
                self.assertEqual(expected[i][j], result[i][j])

    def test_nothing(self):
        result = Solution().levelOrder(None)
        self.assertEqual(0, len(result))

File: algorithms/code/leetcode/lc102_binary_tree_level_order_traversal/lc102_binary_tree_level_order_traversal.rs
// https://leetcode.com/problems/binary-tree-level-order-traversal/
// #medium

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Rc<RefCell<Self>> {
        Rc::new(RefCell::new(TreeNode {
            val,
            left: None,
            right: None,
        }))
    }
}

struct AnnotatedNode {
    node: Rc<RefCell<TreeNode>>,
    depth: i32,
}

pub struct Solution;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut queue: Vec<AnnotatedNode> = Vec::new();
        if let Some(node) = root {
            queue.push(AnnotatedNode { node, depth: 1 });
        }
        let mut depth = 0;
        while !queue.is_empty() {
            let annotated = queue.remove(0);
            if depth != annotated.depth {
                depth = annotated.depth;
                result.push(Vec::new());
            }
            let annotated_node = annotated.node.borrow();
            if let Some(line) = result.last_mut() {
                line.push(annotated_node.val);
            }
            if let Some(left) = &annotated_node.left {
                queue.push(AnnotatedNode {
                    node: Rc::clone(left),
                    depth: annotated.depth + 1,
                });
            }
            if let Some(right) = &annotated_node.right {
                queue.push(AnnotatedNode {
                    node: Rc::clone(right),
                    depth: annotated.depth + 1,
                });
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let n3 = TreeNode::new(3);
        let n7 = TreeNode::new(7);
        let n9 = TreeNode::new(9);
        let n15 = TreeNode::new(15);
        let n20 = TreeNode::new(20);
        n3.borrow_mut().left = Some(Rc::clone(&n9));
        n3.borrow_mut().right = Some(Rc::clone(&n20));
        n20.borrow_mut().left = Some(Rc::clone(&n15));
        n20.borrow_mut().right = Some(Rc::clone(&n7));
        let expected = vec![vec![3], vec![9, 20], vec![15, 7]];
        let result = Solution::level_order(Some(n3));
        assert_eq!(expected, result);
    }

    #[test]
    fn test_nothing() {
        let result = Solution::level_order(None);
        assert!(result.is_empty());
    }
}
